### ISSUE TITLE
Modification after `embark-export` doesn't work

### ISSUE THREAD

**Issue by iyefrat** [oantolin/embark#226]
I want to make an `embark-wgrep` command, which just does `embark-export` on the correct type of buffer and then does `wgrep-change-to-wgrep`, but it seems that `(embark-export)` terminates the function containing it.

To reproduce, run `emacs -Q` and evaluate the following:

```emacs
(require 'package)
(add-to-list 'package-archives '("melpa" . "https://melpa.org/packages/") t)
(package-initialize)
(package-refresh-contents)
(package-install 'embark)
(package-install 'wgrep)
(package-install 'selectrum)
(package-install 'consult)
(package-install 'embark-consult)
(require 'embark)
(require 'wgrep)
(require 'consult)
(require 'embark-consult)
(selectrum-mode +1)

(defun embark-wgrep ()
  "Invoke a wgrep buffer on the current selectrum results, if supported."
  (interactive)
  (require 'wgrep)
  (pcase-let ((`(,type . ,candidates)
               (run-hook-with-args-until-success 'embark-candidate-collectors)))
      (if (not (eq type 'consult-grep))
          (user-error "embark category %S doesn't support wgrep" type)
        (embark-export)
        (wgrep-change-to-wgrep-mode))))
        
(defun embark-export-foo ()
  (interactive)
        (embark-export)
        (message "foo"))
        
(global-set-key (kbd "C-c C-e") #'embark-wgrep)
(global-set-key (kbd "C-c C-f") #'embark-export-foo)
```
Then try either `C-c C-e` on a `consult-grep` search, and see that `wgrep-mode` has not been activated, or `C-c C-f` on any exportable buffer, and see that "foo" doesn't show up in messages.

**Comment by oantolin** [comment]
Yeah. The main use case for export is when the candidates come from the minibuffer, in which case the command that opened said minibuffer is supposed to be canceled, so yes aborting the command is by design. Use `(run-at-time 0 nil #'wgrep-change-to-wgrep-mode)` _before_ calling export to schedule the change to wgrep mode.

**Comment by oantolin** [comment]
(The `run-at-time` can also pass parameters to its callback, so you can rewrite your `embsrk-export-foo` function using `(run-at-time 0 nil #'message "foo")`.)

**Comment by iyefrat** [comment]
This works for the `embark-export-foo` case, but not for `embark-wgrep`. The echo area indicates that `wgrep-mode` is on with the "Press `C-x C-s`..." message, but the buffer isn't editable, and from checking the available keybindings it doesn't seem that the wgrep commands (e.g. `C-x C-s`) have actually been loaded.

**Comment by oantolin** [comment]
Ah, the wgrep-change-to-wgrep-mode command probably ran in the wrong buffer. Maybe I should add an embark-after-export-hook. In the meantime you might be able to figure out how to run it in the newly created buffer. (I'm not at the computer right now, but will experiment later).

**Comment by oantolin** [comment]
It could also be that wgrep-change-to-wgrep-mode is running too soon. I hope it's not that.

**Comment by minad** [comment]
An embark-after-export-hook sounds like a good idea! Would something be needed for collect-snapshot, collect-live too, probably it is sufficient to hook into the embark-collect-mode?

**Comment by iyefrat** [comment]
> It could also be that wgrep-change-to-wgrep-mode is running too soon. I hope it's not that.

I don't think this is it, running it with `(run-at-time 5 ...)` doesn't work either.

**Comment by oantolin** [comment]
> Would something be needed for collect-snapshot, collect-live too, probably it is sufficient to hook into the embark-collect-mode?

Those are OK, I believe. There is the embark-collect-mode-hook as you say, and there is also an after revert hook.

**Comment by oantolin** [comment]
I added an `embark-after-export-hook` that runs in the exported buffer. Now you can rewrite your command as follows:

```elisp
(defun embark-wgrep ()
  "Invoke a wgrep buffer on the current selectrum results, if supported."
  (interactive)
  (require 'wgrep)
  (pcase-let ((`(,type . ,candidates)
               (run-hook-with-args-until-success 'embark-candidate-collectors)))
      (if (not (eq type 'consult-grep))
          (user-error "embark category %S doesn't support wgrep" type)
        (let ((embark-after-export-hook #'wgrep-change-to-wgrep-mode))
          (embark-export)))))
```
